libs/wingc/src/type_check_assert.rs

```diff
@@ -1,12 +1,14 @@
 use crate::{
 	ast::{Expr, Scope},
-	type_check::Types,
+	diagnostic::{report_diagnostic, Diagnostic},
+	type_check::{JsonData, JsonDataKind, Type, Types},
 	visit::{self, Visit},
 };
 
 /// This visitor validates:
 /// 1. All expressions in the AST were resolved to some type.
 /// 2. If any were resolved to `Unresolved` then we should have generated some error diagnostics.
+/// 3. Any Json literal expression only has legal json values unless it is being used as a struct.
```

Comment:
Why is this check added here instead of during the main type checking phase? This code smells off to me because it seems like `TypeCheckAssert` is no longer having just a single responsibility.

Reply:
Is this unique to Json? How is this different from inferring the type of an Array:

```js
let foo = () => {
  return "hi";
};
let bar = foo();

let x = [bar];
```

(BTW this example throws the error `Expected type to be "unknown", but got "unknown" instead`... smells like a bug)

Reply:
Json-legalness is not really part of our types, so inference can't known whether or not a given inferred type will be json legal or not. 
If I make an array `let x = [bar]` but I don't know what `bar` is (yet), nothing unsound will be introduced so you don't need to go back and double check it. but if it's `let x = Json [bar]` then there are limitations on what `bar` is allowed to be.

Additionally, the added feature of casting Json to structs is unique and requires this check to happen later. Consider this example:

```
bring cloud;

struct HasBucket {
  b: cloud.Bucket;
}

let regularJson = {
  b: new cloud.Bucket()
//   ^^^^^^^^^^^^^^^^^^ This is an error
};

let hasBucket: HasBucket = {
  b: new cloud.Bucket()
//.  ^^^^^^^^^^^^^^^^^^ This is not an error because the literal is only used in the context of being casted to a struct
};

let hasBucket2: HasBucket = regularJson;
//                          ^^^^^^^^^^^ This is not an error because there's no need to double-report the issue
```

(worth noting I just found a bug writing out this example, but I'll fix it now and add this example)

---

> BTW this example throws the error Expected type to be "unknown", but got "unknown" instead... smells like a bug

You are right, there is an issue with inference. I'll look into a fix and add that as a test case.

Reply:
Thanks for the explanation 👍  - very helpful.